File: 3885-count-special-triplets/count-special-triplets.java
class Solution {
    public int specialTriplets(int[] nums) {
        HashMap<Integer,Long> mp1 = new HashMap<>();
        HashMap<Integer,Long> mp2 = new HashMap<>();

        int n =  nums.length;

        for(int i=n-1;i>=1;i--)
        {
            mp2.merge(nums[i],1L,Long::sum);
        }
        mp1.put(nums[0],1L);

        long cnt=0;
        //System.out.println(mp2.get(0));
        for(int j=1;j<n;j++)
        {
            mp2.put(nums[j],mp2.get(nums[j])-1L);
            cnt+=mp1.getOrDefault( 2*nums[j], 0L) * mp2.getOrDefault( 2*nums[j], 0L);
            mp1.put(nums[j], mp1.getOrDefault( nums[j], 0L) + 1L);
        }
        cnt=cnt%((long)1e9+7);
        return (int)cnt;
    }
}

File: 3885-count-special-triplets/count-special-triplets.cpp
class Solution {
public:
    int specialTriplets(vector<int>& nums) {
        map<int,long long> mp1, mp2;
        int n=nums.size();
        long long cnt=0;

        for(int i=n-1;i>=1;i--)mp2[nums[i]]++;

        mp1[nums[0]]++;

        for(int i=1;i<n-1;i++)
        {
            mp2[nums[i]]--;
            cnt+=mp1[2*nums[i]]*mp2[2*nums[i]];
            mp1[nums[i]]++;
        }
        cnt%=(long long)1e9+7;
        return (int)cnt;
    }
};